contracts/Router.sol

```diff
@@ -98,21 +104,14 @@ contract Router is IRouter, EIP712, FeeGenerator {
     /// @param user The user address
     /// @return The calculated agent address for the user
     function calcAgent(address user) external view returns (address) {
-        address result = address(
-            uint160(
-                uint256(
-                    keccak256(
-                        abi.encodePacked(
-                            bytes1(0xff),
-                            address(this),
-                            bytes32(bytes20(user)),
-                            keccak256(abi.encodePacked(type(Agent).creationCode, abi.encode(agentImplementation)))
-                        )
-                    )
-                )
-            )
+        // https://github.com/matter-labs/v2-testnet-contracts/blob/main/l2/contracts/L2ContractHelper.sol
+        bytes32 constructorInputHash = keccak256(abi.encode(agentImplementation));
+
+        bytes32 hash = keccak256(
+            bytes.concat(keccak256("zksyncCreate2"), bytes32(uint256(uint160(address(this)))), bytes32(bytes20(user)), agentBytecodeHash, constructorInputHash)
         );
-        return result;
+
+        return address(uint160(uint256(hash)));
```

Comment:
```suggestion
        return address(bytes20(hash));
```

Reply:
nice! I was thinking if I need to keep it to be the same as [L2ContractHelper](https://github.com/matter-labs/v2-testnet-contracts/blob/b8449bf9c819098cc8bfee0549ff5094456be51d/l2/contracts/L2ContractHelper.sol#L41C31-L41C32). But I already use bytes20 at `bytes32(bytes20(user) `, perhaps we can reuse it here.

Reply:
I found we need to use `address(uint160(uint256(hash))` or the address will be different due to ordering (bytesXX vs uintXX).

Reply:
Ah, I see. Thank you for the clarification.